File: flow_sdk/builtin/cli_workers/claude_cli.py
"""ClaudeCliOptions — builds Claude Code CLI shell command strings.

Extends WorkerCLIOptions with all Claude-specific switches:
session/resume, fork, model, debug, permissions, chrome, worktree, agents.

Auto-injects CLAUDE_PROJECT_DIR from workdir.
"""

from __future__ import annotations

import json
from typing import Any

from flow_sdk.builtin.cli_workers.base import WorkerCLIOptions


class ClaudeCliOptions(WorkerCLIOptions):
    """Builds a ``claude`` CLI shell command string for PTY injection.

    All fields map 1-to-1 to CLI flags. The command object is constructed
    at process creation time (or upsert time) with the intent baked in;
    ``open()`` adds runtime env vars via ``add_env()`` before calling
    ``to_shell_string()``.

    Example::

        cmd = ClaudeCliOptions(session_id="abc-123", resume=True, workdir="/proj")
        cmd.add_env("FLOWPAD_EXECUTION_SCOPE", scope_json)
        shell_str = cmd.to_shell_string()
        # → cd '/proj' && CLAUDE_PROJECT_DIR='/proj' FLOWPAD_EXECUTION_SCOPE='...'
        #   claude --dangerously-skip-permissions --resume 'abc-123'

    Fork example::

        cmd = ClaudeCliOptions(
            session_id="new-uuid",      # new worker_session_id
            resume=True,
            fork_session_id="src-uuid", # the session being forked from
        )
        # → claude ... --resume 'src-uuid' --fork-session --session-id 'new-uuid'
    """

    def __init__(
        self,
        session_id: str | None = None,
        resume: bool = False,
        fork_session_id: str | None = None,
        model: str | None = None,
        debug: bool = False,
        permission_mode: str = "bypassPermissions",
        chrome: bool = False,
        worktree: bool = False,
        agents_json: dict | None = None,
        workdir: str | None = None,
        env_vars: dict[str, str] | None = None,
        print_mode: bool = False,
        add_dirs: list[str] | None = None,
    ) -> None:
        super().__init__(workdir=workdir, env_vars=env_vars)
        self.session_id = session_id
        self.resume = resume
        self.fork_session_id = fork_session_id
        self.model = model
        self.debug = debug
        self.permission_mode = permission_mode
        self.chrome = chrome
        self.worktree = worktree
        self.agents_json = agents_json
        self.print_mode = print_mode
        self.add_dirs: list[str] = list(add_dirs or [])

        # Auto-inject CLAUDE_PROJECT_DIR from workdir
        if workdir:
            self.env_vars.setdefault("CLAUDE_PROJECT_DIR", workdir)

    # ------------------------------------------------------------------
    # WorkerCLIOptions contract
    # ------------------------------------------------------------------

    def _build_worker_args(self) -> list[str]:
        import shlex

        args: list[str] = ["claude"]

        if self.permission_mode == "bypassPermissions":
            args.append("--dangerously-skip-permissions")
        if self.chrome:
            args.append("--chrome")
        if self.debug:
            args.append("--debug")
        if self.worktree:
            args.append("--worktree")

        if self.resume and self.session_id:
            if self.fork_session_id:
                # Fork: --resume <source> --fork-session --session-id <new>
                args.append(f"--resume {shlex.quote(self.fork_session_id)}")
                args.append("--fork-session")
                args.append(f"--session-id {shlex.quote(self.session_id)}")
            else:
                args.append(f"--resume {shlex.quote(self.session_id)}")
        elif self.session_id:
            args.append(f"--session-id {shlex.quote(self.session_id)}")

        if self.model:
            args.append(f"--model {shlex.quote(self.model)}")
        if self.agents_json:
            args.append(f"--agents {shlex.quote(json.dumps(self.agents_json))}")
        for d in self.add_dirs:
            args.append(f"--add-dir {shlex.quote(d)}")

        if self.print_mode:
            args.append("-p")

        return args

    def to_shell_string(self, instruction: str | None = None) -> str:
        """Build shell command with --add-dir flags appended after the instruction.

        Claude CLI requires that the prompt/instruction appear before --add-dir
        flags when both are present. This override separates --add-dir from the
        other args so _build_posix can insert the instruction first.
        """
        import sys

        args = self._build_worker_args()
        if sys.platform == "win32":
            return self._build_win32(args, instruction)

        main_args = [a for a in args if not a.startswith("--add-dir ")]
        add_dir_args = [a for a in args if a.startswith("--add-dir ")]

        cmd = self._build_posix(main_args, instruction)
        if add_dir_args:
            cmd += " " + " ".join(add_dir_args)
        return cmd

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_json(self) -> dict[str, Any]:
        d = super().to_json()
        d.update({
            "worker_type": "claude",
            "session_id": self.session_id,
            "resume": self.resume,
            "fork_session_id": self.fork_session_id,
            "model": self.model,
            "debug": self.debug,
            "permission_mode": self.permission_mode,
            "chrome": self.chrome,
            "worktree": self.worktree,
            "agents_json": self.agents_json,
            "print_mode": self.print_mode,
            "add_dirs": self.add_dirs,
        })
        return d

    def to_spawn_args(self, instruction: str | None = None) -> tuple[list[str], dict[str, str]]:
        """Build argv list and env dict for PtyProcess.spawn() — no shell intermediary.

        Builds argv directly (each flag and value as separate elements) rather than
        going through shell-string quoting.
        """
        argv: list[str] = ["claude"]

        if self.permission_mode == "bypassPermissions":
            argv.append("--dangerously-skip-permissions")
        if self.chrome:
            argv.append("--chrome")
        if self.debug:
            argv.append("--debug")
        if self.worktree:
            argv.append("--worktree")

        if self.resume and self.session_id:
            if self.fork_session_id:
                argv.extend(["--resume", self.fork_session_id, "--fork-session", "--session-id", self.session_id])
            else:
                argv.extend(["--resume", self.session_id])
        elif self.session_id:
            argv.extend(["--session-id", self.session_id])

        if self.model:
            argv.extend(["--model", self.model])
        if self.agents_json:
            import json as _json
            argv.extend(["--agents", _json.dumps(self.agents_json)])
        if self.print_mode:
            argv.append("-p")

        if instruction:
            argv.extend(["--", instruction])

        for d in self.add_dirs:
            argv.extend(["--add-dir", d])

        return argv, dict(self.env_vars)

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> "ClaudeCliOptions":
        return cls(
            session_id=data.get("session_id"),
            resume=bool(data.get("resume", False)),
            fork_session_id=data.get("fork_session_id"),
            model=data.get("model"),
            debug=bool(data.get("debug", False)),
            permission_mode=data.get("permission_mode", "bypassPermissions"),
            chrome=bool(data.get("chrome", False)),
            worktree=bool(data.get("worktree", False)),
            agents_json=data.get("agents_json"),
            workdir=data.get("workdir"),
            env_vars=data.get("env_vars") or {},
            print_mode=bool(data.get("print_mode", False)),
            add_dirs=list(data.get("add_dirs") or []),
        )

File: ts_sdk/src/cli_workers/claude-cli.ts
/**
 * ClaudeCliOptions — builds Claude Code CLI shell command strings.
 *
 * Mirrors Python flow_sdk/builtin/cli_workers/claude_cli.py exactly.
 * All fields map 1-to-1 to CLI flags. Snake_case field names match
 * the serialized cli_config dict from the backend.
 */

import { WorkerCliOptions, shellQuote } from './base'

const CLAUDE_MODEL_TIERS: Record<string, string> = {
  sm: 'haiku',
  md: 'sonnet',
  lg: 'opus',
}

export interface ClaudeCliOptionsOptions {
  session_id?: string | null
  resume?: boolean
  fork_session_id?: string | null
  model?: string | null
  debug?: boolean
  permission_mode?: string
  chrome?: boolean
  worktree?: boolean
  agents_json?: Record<string, any> | null
  workdir?: string
  env_vars?: Record<string, string>
  add_dirs?: string[]
  print_mode?: boolean
  /** One of "text" | "json" | "stream-json"; omit for CLI default. */
  output_format?: string | null
  /** Required when output_format === "stream-json"; auto-enabled in that case. */
  verbose?: boolean
}

export class ClaudeCliOptions extends WorkerCliOptions {
  session_id?: string | null
  resume: boolean
  fork_session_id?: string | null
  model?: string | null
  debug: boolean
  permission_mode: string
  chrome: boolean
  worktree: boolean
  agents_json?: Record<string, any> | null
  addDirs: string[]
  print_mode: boolean
  output_format?: string | null
  verbose: boolean

  constructor(opts: ClaudeCliOptionsOptions = {}) {
    super(opts.workdir ?? undefined, opts.env_vars)
    this.session_id = opts.session_id ?? undefined
    this.resume = opts.resume ?? false
    this.fork_session_id = opts.fork_session_id ?? undefined
    this.model = opts.model ?? undefined
    this.debug = opts.debug ?? true
    this.permission_mode = opts.permission_mode ?? 'bypassPermissions'
    this.chrome = opts.chrome ?? false
    this.worktree = opts.worktree ?? false
    this.agents_json = opts.agents_json ?? undefined
    this.addDirs = opts.add_dirs ?? []
    this.print_mode = opts.print_mode ?? false
    this.output_format = opts.output_format ?? undefined
    this.verbose = (opts.verbose ?? false) || opts.output_format === 'stream-json'

    // Auto-inject CLAUDE_PROJECT_DIR from workdir (same as Python setdefault)
    if (opts.workdir) {
      this.envVars['CLAUDE_PROJECT_DIR'] ??= opts.workdir
    }
  }

  get resolvedModel(): string | undefined {
    return this.model ? (CLAUDE_MODEL_TIERS[this.model] ?? this.model) : undefined
  }

  protected _buildWorkerArgs(): string[] {
    const args: string[] = ['claude']

    if (this.permission_mode === 'bypassPermissions') {
      args.push('--dangerously-skip-permissions')
    }
    if (this.chrome) args.push('--chrome')
    if (this.debug) args.push('--debug')
    if (this.worktree) args.push('--worktree')
    if (this.verbose) args.push('--verbose')
    if (this.output_format) args.push(`--output-format ${shellQuote(this.output_format)}`)

    if (this.resume && this.session_id) {
      args.push(`--resume ${shellQuote(this.session_id)}`)
      if (this.fork_session_id) {
        args.push('--fork-session')
        args.push(`--session-id ${shellQuote(this.fork_session_id)}`)
      }
    } else if (this.session_id) {
      args.push(`--session-id ${shellQuote(this.session_id)}`)
    }

    if (this.resolvedModel) args.push(`--model ${shellQuote(this.resolvedModel)}`)
    if (this.agents_json) args.push(`--agents ${shellQuote(JSON.stringify(this.agents_json))}`)
    for (const d of this.addDirs) {
      args.push(`--add-dir ${shellQuote(d)}`)
    }
    if (this.print_mode) args.push('-p')

    return args
  }

  toJson(): Record<string, any> {
    return {
      ...super.toJson(),
      worker_type: 'claude',
      session_id: this.session_id,
      resume: this.resume,
      fork_session_id: this.fork_session_id,
      model: this.model,
      debug: this.debug,
      permission_mode: this.permission_mode,
      chrome: this.chrome,
      worktree: this.worktree,
      agents_json: this.agents_json,
      add_dirs: this.addDirs,
      print_mode: this.print_mode,
      output_format: this.output_format,
      verbose: this.verbose,
    }
  }

  static fromJson(data: Record<string, any>): ClaudeCliOptions {
    return new ClaudeCliOptions({
      session_id: data['session_id'] ?? undefined,
      resume: Boolean(data['resume'] ?? false),
      fork_session_id: data['fork_session_id'] ?? undefined,
      model: data['model'] ?? undefined,
      debug: data['debug'] !== undefined ? Boolean(data['debug']) : true,
      permission_mode: data['permission_mode'] ?? 'bypassPermissions',
      chrome: Boolean(data['chrome'] ?? false),
      worktree: Boolean(data['worktree'] ?? false),
      agents_json: data['agents_json'] ?? undefined,
      workdir: data['workdir'] ?? undefined,
      env_vars: data['env_vars'] ?? {},
      add_dirs: data['add_dirs'] ?? [],
      print_mode: Boolean(data['print_mode'] ?? false),
      output_format: data['output_format'] ?? undefined,
      verbose: Boolean(data['verbose'] ?? false),
    })
  }
}
